### PR TITLE
DAGMC CAD-based geometry as universes

### PR DESCRIPTION
**Pull request by pshriwise**
We currently only support DAGMC models where the entire geometry is represented as CAD-based tesselations. This PR adds a DAGMC Universe derived from the standard CSG universe. This allows one to construct geometries from both CAD and CSG. This enables repeated CAD geometries in OpenMC like this lattice of DAGMC pincells from our regression tests

dag_lattice

Or, as a more interesting case, one can rotate a section of a tokamak into a set of CSG cells to model the entire device with a smaller DAGMC memory footprint. (Thanks @Shimwell for providing this quarter-tokamak model as a demonstration.)

Here is the entire tokamak model in Cubit:

<p style="text-align:center;">

</p>

We can add this model as a fill to the appropriate cells and rotate the model 90 degrees to get this X-Y view:

tokamak_ex

Adding two more cells and rotations gives use the entire device:

Screenshot from 2021-04-27 00-08-30

## Input Changes

These updates do require some changes to the current format when the DAGMC model represents the entire geometry. A `geometry.xml` file is now always required. To represent a geometry composed of a single DAGMC model as we do now, a `geometry.xml` file that gives the same behavior should contain

```XML
<?xml version='1.0' encoding='utf-8'?>
<geometry>
<dagmc id="1" filename="dagmc.h5m"/>
</geometry>
```

where the `id` is the universe ID and the filename is the name of the DAGMC model file. 

**Note that the `filename` parameter remoes the restriction that DAGMC files have the name `dagmc.h5m`.**

## Notable code changes

  - Addition of the `DAGUniverse` class
  - Addition of `Cell::geom_type` and `Surface::geom_type`
  - Removal of the `settings::dagmc` parameter from the CAPI and Python API
  - Added a `Universe::find_cell` method
  - Fewer `#ifdef DAGMC`'s throughout the code thanks to the addition of the `GeometryType` enum

## ID management

One difficulty of mixing CSG and CAD geometry is the unknown ID spaces of the CAD geometry before simulation initialization. The `DAGUniverse` class on the Python side has an option, `auto_geom_ids`, to automatically set cell and surface IDs outside of the CSG cell & surface space to avoid conflicts. If this option is not enabled, then an error will occur if the ID spaces overlap.

A similar option, `auto_mat_ids`, is also available for materials to address the case where some of the DAGMC universes are using UWUW materials (material definitions embedded in the `.h5m` along with the geometry.

### Coincident DAGMC/CSG surfaces

Thanks to our preference for hitting surface at higher levels in the geometry, this is hasn't been a problem in the models I've used for testing.

### Managing a particle's ray history

The ray history is a record of facets hit by DAGMC rays to avoid duplicate hits when coincident with a surface or reflecting from a surface. This history is reset when

  - a particle changes direction or undergoes a collision,
  - a particle is not on a surface,
  - and a particle is terminated.
 
Now that a particle can move directly from one DAGMC universe to another, another case is added for

  - a particle exiting the current `DAGUniverse`

For more details on this implementation, please see [this google doc](https://docs.google.com/document/d/1Jbg-8VZhxBXTMxbz1r85GgaFgLBYIlkXbvvXLUhV_K0/edit?usp=sharing) w/ the proposal for this work.

### CODE REVIEW

**Comment by makeclean** [comment]
Ooh very exciting :)
________________________________

**Comment by gridley** [comment]
In case anyone else is wondering, this seems to add pretty much zero overhead to non-DAGMC models. Seems like a really useful upgrade.

**Comment by pshriwise** [comment]
Sorry for the initial CI failures here. Needed to make a correction to the properties of the `DAGUniverse` Python class. This should be good to try out now!

**Comment by Shimwell** [comment]
@pshriwise is it worth me adding a rotation to this example https://github.com/openmc-dev/openmc/pull/1800 . I assume that I just remove the reflecting surface (perhaps using the new dagmc_remove_tags package) and rotate 90, 180 and 270 degrees to make the complete geometry

**Comment by pshriwise** [comment]
> @pshriwise is it worth me adding a rotation to this example #1800 . I assume that I just remove the reflecting surface (perhaps using the new [dagmc_remove_tags](https://github.com/svalinn/remove_dagmc_tags) package) and rotate 90, 180 and 270 degrees to make the complete geometry

I'd welcome it! It would make #1800 dependent on this PR though. I'll leave it up to you as to whether or not you're ok with waiting on these changes to get that example in.

File: src/dagmc.cpp
Comment (makeclean) [inline src/dagmc.cpp:76]: will `u->id_` always be sorted smallest to largest?

File: src/dagmc.cpp
Comment (makeclean) [inline src/dagmc.cpp:155]: say i tally in cell 3, which is a dagmc cell, do we accumulate tally scores across all the instances of that cell individually, or would I get a score for each 'copy'?

File: src/dagmc.cpp
Comment (makeclean) [inline src/dagmc.cpp:235]: Do you need a graveyard even in a universe based model? Do OpenMC cells not bound this?

File: src/dagmc.cpp
Comment (makeclean) [inline src/dagmc.cpp:259]: is this sort of thing not handled by DMD? I thought it dealt with synonyms?

File: src/dagmc.cpp
Comment (makeclean) [inline src/dagmc.cpp:315]: quite comment light, i presume you've making a list like 1-100, 101, 105-109?

File: src/dagmc.cpp
Comment (makeclean) [inline src/dagmc.cpp:248]: maybe a note here - reminder dagmc indices are 1 based

File: tests/regression_tests/rotation/test.py
Comment (makeclean) [inline tests/regression_tests/rotation/test.py:1]: was this file changed on purpose? seems unrelated to the DAGMC based changes? maybe im wrong?

**Review by makeclean — CHANGES_REQUESTED** [review]
Looks mostly good here, I think a few extra new lines, some extra comments and its good :)

**Comment by Shimwell** [comment]
Looks like code block 17 from the cad-based-geometry.ipynb has the old ```settings.dagmc = True``` method  

https://github.com/pshriwise/openmc/blob/5b20fbbf6a377b16d201a34c5cf65471662ac7cc/examples/jupyter/cad-based-geometry.ipynb#L524

File: include/openmc/surface.h
Comment (pshriwise) [reply include/openmc/surface.h:33]: Nope. Definitely not. Unsure as to how that came about. Thanks!

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:76]: It isn't a requirement for `model::universes`, no. Are you concerned the next ID isn't correct?

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:155]: When using a `CellFilter` you'll get the accumulated score over all the instances of that cell in the geometry. If you use a `CellInstanceFilter` filter or a `DistribCellFilter` you can get for each instance of the cell in the geometry.

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:235]: The only case where you now need one is if the DAGMC model is the root universe of the Geometry. I'll add some detection for that so this warning doesn't appear when it shouldn't. Thanks!

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:259]: I believe synonyms are for the property type (`mat:`, `surf:`, `rho:`) as opposed to the value. The logic here is matches the surface BC logic [here](https://github.com/openmc-dev/openmc/blob/4399645859ce5459828721f12fbcaf052a63653a/src/surface.cpp#L137). I thought it made sense for these to be consistent.

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:315]: Correct. I'll add some more explanation here.

File: tests/regression_tests/rotation/test.py
Comment (pshriwise) [reply tests/regression_tests/rotation/test.py:1]: Ah, no you're right. This is the result of a mishandled rebase conflict. I'll sort it out.

File: src/dagmc.cpp
Comment (makeclean) [reply src/dagmc.cpp:76]: just wondering if it leads to any issues, like does it matter if the order isnt monotonically increasing?

**Comment by Shimwell** [comment]
I've been using the branch on an ARC reactor model and have reproduced the example pictures with a different geometry (still 90 degree rotated 4 times).

I've found the ability to turn of the auto assign id numbers useful
It is nice to have a geometry.xml file when using DAGMC
It is great that this branch allows flexability with the dagmc h5m file (previously only dagmc.h5m was accepted).

Actually this feature really saved me and was exactly what I needed to get the simulation results.

Perhaps this is unrelated to this PR but one thing I noticed was that the terminal prints out a lot of (perhaps too much) information when plotting an DAGMC geometry using this method

```
import os
# makes the 3d "cube" style geometry
vox_plot = openmc.Plot()
vox_plot.type = 'voxel'
vox_plot.width = (1700., 1700., 1800.)
vox_plot.pixels = (300, 300, 300)
vox_plot.filename = 'plot_3d_tokamak'
vox_plot.color_by = 'material'
plots = openmc.Plots([vox_plot])
plots.export_to_xml()
openmc.plot_geometry()
os.system('openmc-voxel-to-vtk plot_3d_tokamak.h5 -o plot_3d_tokamak.vti')
```

**Comment by pshriwise** [comment]
> import os
> # makes the 3d "cube" style geometry
> vox_plot = openmc.Plot()
> vox_plot.type = 'voxel'
> vox_plot.width = (1700., 1700., 1800.)
> vox_plot.pixels = (300, 300, 300)
> vox_plot.filename = 'plot_3d_tokamak'
> vox_plot.color_by = 'material'
> plots = openmc.Plots([vox_plot])
> plots.export_to_xml()
> openmc.plot_geometry()
> os.system('openmc-voxel-to-vtk plot_3d_tokamak.h5 -o plot_3d_tokamak.vti')

Yeah, I'm seeing that too. Thanks for sharing this. Appears to be a warning about a corner case in MOAB's triangle intersection routine. I'll have to look into that separately.

**Comment by pshriwise** [comment]
> Looks like code block 17 from the cad-based-geometry.ipynb has the old `settings.dagmc = True` method

Thanks! Just took care of that one.

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:76]: The IDs can be in any order here. Whenever we find one that's bigger, we update `next_univ_id`. If we happened to find the largest ID first, we'd only update that variable once.

**Comment by pshriwise** [comment]
Got this updated after the merge of #1834.

**Review by Shimwell — APPROVED** [review]
I've been using this branch for a while now and not noticed any problems. My C++ skills are not really good enough to understand all the changes. But from a user point of view the docs, examples and usage is all fine.

File: openmc/universe.py
Comment (paulromano) [inline openmc/universe.py:683]: Defining `fmt_str` upfront actually doesn't save any work. In the case where you repeat the literal twice, Python's compiler is smart enough to recognize that it is the same object. By setting it as a various like this, you are actually giving the interpreter more work to do.
```suggestion
        string = super().__repr__()
        string += '{: <16}=\t{}\n'.format('\tGeom', 'DAGMC')
        string += '{: <16}=\t{}\n'.format('\tFile', self.filename)
```

File: openmc/universe.py
Comment (paulromano) [inline openmc/universe.py:714]: Should this be a warning or something?

File: src/particle.cpp
Comment (paulromano) [inline src/particle.cpp:389]:
```suggestion
```

File: src/surface.cpp
Comment (paulromano) [inline src/surface.cpp:208]:
```suggestion
  } else if (geom_type_ == GeometryType::CSG) {
```

File: tests/regression_tests/dagmc/legacy/test.py
Comment (paulromano) [inline tests/regression_tests/dagmc/legacy/test.py:31]:
```suggestion
    model.geometry = openmc.Geometry(dag_univ)
```

File: tests/regression_tests/dagmc/refl/test.py
Comment (paulromano) [inline tests/regression_tests/dagmc/refl/test.py:32]:
```suggestion
        model.geometry = openmc.Geometry(dag_univ)
```

File: openmc/universe.py
Comment (paulromano) [inline openmc/universe.py:730]: Maybe make this 'dagmc_universe'?

File: tests/regression_tests/rotation/inputs_true.dat
Comment (paulromano) [inline tests/regression_tests/rotation/inputs_true.dat:1]: Doesn't look like this file is used? The rotation test is just a normal (non Python API) test.

File: tests/regression_tests/dagmc/universes/test.py
Comment (paulromano) [inline tests/regression_tests/dagmc/universes/test.py:56]: Might be more exciting to put this DAGMC pincell in a lattice so that it appears multiple times?

File: tests/regression_tests/dagmc/universes/test.py
Comment (paulromano) [inline tests/regression_tests/dagmc/universes/test.py:58]:
```suggestion
        model.geometry = openmc.Geometry([bounding_cell])
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Really awesome feature @pshriwise! Looking forward to having this as part of our next release.

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp:76]: Are you happy with this implementation @makeclean?

File: include/openmc/cell.h
Comment (pshriwise) [reply include/openmc/cell.h:124]: For sure. Thanks!

File: include/openmc/dagmc.h
Comment (pshriwise) [reply include/openmc/dagmc.h:110]: Made private. Updated constructor to take those as parameters. Added accessors.

File: openmc/universe.py
Comment (pshriwise) [reply openmc/universe.py:714]: I need to think about this a bit more. I don't see any reason why we can't clone a `DAGMCUniverse`. I wouldn't necessarily recommend it, but it should be possible.

File: tests/regression_tests/rotation/inputs_true.dat
Comment (pshriwise) [reply tests/regression_tests/rotation/inputs_true.dat:1]: Thanks! looks like this was leftover from my update to the new `develop` history after getting rid of that `plot_settings.pkl` file.

File: src/dagmc.cpp
Comment (pshriwise) [reply src/dagmc.cpp]: Reorganized the code a bit so we can perform this check in 302f12b5ffba88c78ce2cb39722e76ade9672735. Hopefully this is what you had in mind!

File: tests/regression_tests/dagmc/universes/test.py
Comment (pshriwise) [reply tests/regression_tests/dagmc/universes/test.py]: Updated
 in 263addf

**Comment by paulromano** [comment]
@pshriwise Thanks for incorporating all the feedback. I have a few more requests:
1. The new `<dagmc_universe>` element needs to be documented in `docs/source/io_formats/geometry.rst`
2. The `geom_type` attribute for surfaces needs to be documented in `docs/source/io_formats/summary.rst`
3. It looks like a geometry that has a DAGMC universe does not "round trip" correctly to and from XML. That is, if you `geom.export_to_xml()` and then `new_geom = openmc.Geometry.from_xml(...); new_geom.export_to_xml('new_geom.xml')`, the `new_geom.xml` file is not the same as the original.

File: openmc/surface.py
Comment (pshriwise) [reply openmc/surface.py]: Addressed this in f2e0af3.

**Comment by pshriwise** [comment]
> @pshriwise Thanks for incorporating all the feedback. I have a few more requests:
> 
> 1. The new `<dagmc_universe>` element needs to be documented in `docs/source/io_formats/geometry.rst`
> 2. The `geom_type` attribute for surfaces needs to be documented in `docs/source/io_formats/summary.rst`
> 3. It looks like a geometry that has a DAGMC universe does not "round trip" correctly to and from XML. That is, if you `geom.export_to_xml()` and then `new_geom = openmc.Geometry.from_xml(...); new_geom.export_to_xml('new_geom.xml')`, the `new_geom.xml` file is not the same as the original.

Thanks for  @paulromano! Docs have all been updated.

I've updated the export/import code for XML so that it handles DAGMC universes properly and can reproduce the same `geometry.xml` reliably. I also decided that we should really be able to reproduce the `geometry.xml` from a `summary.h5` file as well, so I've added a `DAGUniverse::to_hdf5` method that writes out all of the necessary information to reproduce the Python object (and corresponding XML element if desired). I've updated the `summary.h5` documentation to reflect these changes as well.

I noticed along the way that surface names weren't being picked up in `openmc.Surface.from_xml_element`, so I took care of that in 25cf7a3e9712f80b05e918311fbd5e65cfd6d627.

File: docs/source/io_formats/geometry.rst
Comment (paulromano) [inline docs/source/io_formats/geometry.rst:386]:
```suggestion
    *Default*: None
```

File: docs/source/io_formats/geometry.rst
Comment (paulromano) [inline docs/source/io_formats/geometry.rst:392]:
```suggestion
    *Default*: None
```

File: docs/source/io_formats/geometry.rst
Comment (paulromano) [inline docs/source/io_formats/geometry.rst:390]: As we discussed offline, this limit is probably historical and no longer applies (unless you've found something that indicates we are indeed limited to 52 characters)

File: docs/source/io_formats/geometry.rst
Comment (paulromano) [inline docs/source/io_formats/geometry.rst:413]:
```suggestion
```

File: openmc/universe.py
Comment (paulromano) [inline openmc/universe.py:761]: For consistency, let's name this `from_xml_element`

File: openmc/universe.py
Comment (paulromano) [inline openmc/universe.py:726]: Do we really want to allow these universes to be named? Right now, although you can give a regular universe a name, nothing is actually done with it since we don't write a `<universe>` element in the XML. At the very least, if the `name` is not assigned, we shouldn't create an empty `name=""` attribute in the XML.

File: docs/source/io_formats/geometry.rst
Comment (paulromano) [inline docs/source/io_formats/geometry.rst:413]:
```suggestion
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for the latest round of updates. A few more requests based on those changes... we're getting close!

File: openmc/universe.py
Comment (pshriwise) [reply openmc/universe.py:726]: Good point. I was realizing this too looking at the import/export XML methods yesterday. I think it's probably best to make the behavior of the name attribute as similar to the regular universe as possible.

File: docs/source/io_formats/geometry.rst
Comment (pshriwise) [reply docs/source/io_formats/geometry.rst:390]: Nope! I'll go ahead and update other instances of this as well.

File: openmc/universe.py
Comment (pshriwise) [reply openmc/universe.py:726]: One instance where it would be nice to have the universe name is when looking at different levels of the geometry in the plotter so we can display universe names in the status bar. I think it can be dealt with in a separate PR if we feel it's important enough to get those names into the XML files though.

File: openmc/geometry.py
Comment (paulromano) [inline openmc/geometry.py:164]:
```suggestion
            dag_univ = openmc.DAGMCUniverse.from_xml_element(elem)
```

**Review by paulromano — CHANGES_REQUESTED** [review]

**Review by paulromano — APPROVED** [review]

**Comment by paulromano** [comment]
@makeclean Do you have any further comments on this PR?

**Review by makeclean — APPROVED** [review]

**Comment by makeclean** [comment]
LGTM thanks @pshriwise

**Pull request merged**
